File: berrychain/crypto.py
"""
Cryptographic primitives for BerryChain.

* Ed25519 for account signatures
* X25519 + HKDF + ChaCha20-Poly1305 for wrapping packet keys to a buyer
* ChaCha20-Poly1305 for encrypting packet contents
* SHA-256 for hashing, addresses and proof-of-work
"""

from __future__ import annotations

import hashlib
import json
import os
from typing import Any

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)
from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from . import params

ADDRESS_PREFIX = "brry1"
_RAW = serialization.Encoding.Raw
_PUB_RAW = serialization.PublicFormat.Raw
_PRIV_RAW = serialization.PrivateFormat.Raw
_NOENC = serialization.NoEncryption()


# ---------------------------------------------------------------------------
# Hashing / canonical encoding
# ---------------------------------------------------------------------------

def sha256(data: bytes) -> bytes:
    return hashlib.sha256(data).digest()


def sha256d(data: bytes) -> bytes:
    return sha256(sha256(data))


def canonical_json(obj: Any) -> bytes:
    """Deterministic JSON encoding used for every hash and signature."""
    return json.dumps(obj, sort_keys=True, separators=(",", ":"), ensure_ascii=True).encode()


def hash_obj(obj: Any) -> str:
    return sha256(canonical_json(obj)).hex()


# ---------------------------------------------------------------------------
# Signing keys and addresses
# ---------------------------------------------------------------------------

def generate_signing_keypair() -> tuple[str, str]:
    """Return (private_hex, public_hex) for a new Ed25519 key."""
    priv = Ed25519PrivateKey.generate()
    priv_hex = priv.private_bytes(_RAW, _PRIV_RAW, _NOENC).hex()
    pub_hex = priv.public_key().public_bytes(_RAW, _PUB_RAW).hex()
    return priv_hex, pub_hex


def public_from_private(priv_hex: str) -> str:
    priv = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(priv_hex))
    return priv.public_key().public_bytes(_RAW, _PUB_RAW).hex()


def sign(priv_hex: str, message: bytes) -> str:
    priv = Ed25519PrivateKey.from_private_bytes(bytes.fromhex(priv_hex))
    return priv.sign(message).hex()


def verify(pub_hex: str, message: bytes, sig_hex: str) -> bool:
    try:
        pub = Ed25519PublicKey.from_public_bytes(bytes.fromhex(pub_hex))
        pub.verify(bytes.fromhex(sig_hex), message)
        return True
    except (InvalidSignature, ValueError):
        return False


def address_from_pubkey(pub_hex: str) -> str:
    payload = sha256(bytes.fromhex(pub_hex))[:20]
    checksum = sha256(b"berry-address" + payload)[:4]
    return ADDRESS_PREFIX + (payload + checksum).hex()


def is_valid_address(addr: Any) -> bool:
    if addr == params.TREASURY_ADDRESS:
        return True
    if not isinstance(addr, str) or not addr.startswith(ADDRESS_PREFIX):
        return False
    body = addr[len(ADDRESS_PREFIX):]
    if len(body) != 48:
        return False
    try:
        raw = bytes.fromhex(body)
    except ValueError:
        return False
    payload, checksum = raw[:20], raw[20:]
    return sha256(b"berry-address" + payload)[:4] == checksum


# ---------------------------------------------------------------------------
# Encryption keys (X25519) and key wrapping
# ---------------------------------------------------------------------------

def generate_encryption_keypair() -> tuple[str, str]:
    priv = X25519PrivateKey.generate()
    return (
        priv.private_bytes(_RAW, _PRIV_RAW, _NOENC).hex(),
        priv.public_key().public_bytes(_RAW, _PUB_RAW).hex(),
    )


def encryption_public_from_private(priv_hex: str) -> str:
    priv = X25519PrivateKey.from_private_bytes(bytes.fromhex(priv_hex))
    return priv.public_key().public_bytes(_RAW, _PUB_RAW).hex()


def is_valid_enc_pub(pub: Any) -> bool:
    if not isinstance(pub, str) or len(pub) != 64:
        return False
    try:
        X25519PublicKey.from_public_bytes(bytes.fromhex(pub))
        return True
    except ValueError:
        return False


def _derive(shared: bytes, epk: bytes, rpk: bytes) -> bytes:
    return HKDF(algorithm=SHA256(), length=32, salt=epk + rpk, info=b"berry-ecies-v1").derive(shared)


def wrap_to_recipient(recipient_pub_hex: str, plaintext: bytes) -> dict:
    """ECIES: encrypt `plaintext` so only the holder of the recipient key can read it."""
    rpk = X25519PublicKey.from_public_bytes(bytes.fromhex(recipient_pub_hex))
    eph = X25519PrivateKey.generate()
    epk_bytes = eph.public_key().public_bytes(_RAW, _PUB_RAW)
    key = _derive(eph.exchange(rpk), epk_bytes, bytes.fromhex(recipient_pub_hex))
    nonce = os.urandom(12)
    ct = ChaCha20Poly1305(key).encrypt(nonce, plaintext, epk_bytes)
    return {"epk": epk_bytes.hex(), "nonce": nonce.hex(), "ct": ct.hex()}


def unwrap_from_sender(recipient_priv_hex: str, blob: dict) -> bytes:
    priv = X25519PrivateKey.from_private_bytes(bytes.fromhex(recipient_priv_hex))
    rpk_bytes = priv.public_key().public_bytes(_RAW, _PUB_RAW)
    epk_bytes = bytes.fromhex(blob["epk"])
    key = _derive(priv.exchange(X25519PublicKey.from_public_bytes(epk_bytes)), epk_bytes, rpk_bytes)
    return ChaCha20Poly1305(key).decrypt(bytes.fromhex(blob["nonce"]), bytes.fromhex(blob["ct"]), epk_bytes)


# ---------------------------------------------------------------------------
# Packet content encryption
# ---------------------------------------------------------------------------

def new_packet_key() -> bytes:
    return os.urandom(32)


def encrypt_packet(key: bytes, plaintext: bytes) -> bytes:
    nonce = os.urandom(12)
    return nonce + ChaCha20Poly1305(key).encrypt(nonce, plaintext, b"berry-packet-v1")


def decrypt_packet(key: bytes, ciphertext: bytes) -> bytes:
    nonce, ct = ciphertext[:12], ciphertext[12:]
    return ChaCha20Poly1305(key).decrypt(nonce, ct, b"berry-packet-v1")


def key_commitment(key: bytes) -> str:
    """Hash published in a listing; lets a buyer check the delivered key is the listed one."""
    return sha256(b"berry-key-commit" + key).hex()

File: app/lib/core/crypto.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:crypto/crypto.dart' as c;
import 'package:cryptography/cryptography.dart';
import 'package:pointycastle/export.dart' as pc;

import 'canonical.dart';

/// The chain's primitives, same choices as the reference node:
/// Ed25519 signatures, X25519 + HKDF-SHA256 + ChaCha20-Poly1305 key wrapping,
/// ChaCha20-Poly1305 content encryption, SHA-256 everywhere else, scrypt for
/// the wallet file.

const addressPrefix = 'brry1';
const _addressTag = 'berry-address';
const _ecies = 'berry-ecies-v1';
const _packetAad = 'berry-packet-v1';
const _keyCommit = 'berry-key-commit';
const _walletAad = 'berry-wallet-v1';

final _rng = Random.secure();

Uint8List randomBytes(int n) =>
    Uint8List.fromList(List<int>.generate(n, (_) => _rng.nextInt(256)));

String toHex(List<int> b) =>
    b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();

Uint8List fromHex(String h) {
  if (h.length.isOdd) throw const FormatException('odd hex length');
  final out = Uint8List(h.length ~/ 2);
  for (var i = 0; i < out.length; i++) {
    out[i] = int.parse(h.substring(2 * i, 2 * i + 2), radix: 16);
  }
  return out;
}

Uint8List sha256(List<int> data) => Uint8List.fromList(c.sha256.convert(data).bytes);
Uint8List sha256d(List<int> data) => sha256(sha256(data));
String hashObj(Object? obj) => toHex(sha256(canonicalBytes(obj)));

/// Big-endian integer view of a 32-byte digest, for target comparisons.
BigInt bigFromBytes(List<int> b) {
  var r = BigInt.zero;
  for (final x in b) {
    r = (r << 8) | BigInt.from(x);
  }
  return r;
}

// ---------------------------------------------------------------- signing

final _ed = Ed25519();
final _x = X25519();

Future<String> signPublicFromPrivate(String privHex) async {
  final kp = await _ed.newKeyPairFromSeed(fromHex(privHex));
  return toHex((await kp.extractPublicKey()).bytes);
}

Future<String> sign(String privHex, List<int> message) async {
  final kp = await _ed.newKeyPairFromSeed(fromHex(privHex));
  return toHex((await _ed.sign(message, keyPair: kp)).bytes);
}

Future<bool> verify(String pubHex, List<int> message, String sigHex) async {
  try {
    final pub = SimplePublicKey(fromHex(pubHex), type: KeyPairType.ed25519);
    return await _ed.verify(message, signature: Signature(fromHex(sigHex), publicKey: pub));
  } catch (_) {
    return false;
  }
}

String addressFromPubkey(String pubHex) {
  final payload = sha256(fromHex(pubHex)).sublist(0, 20);
  final checksum = sha256([...utf8.encode(_addressTag), ...payload]).sublist(0, 4);
  return addressPrefix + toHex(payload) + toHex(checksum);
}

bool isValidAddress(String? addr) {
  if (addr == null || !addr.startsWith(addressPrefix)) return false;
  final body = addr.substring(addressPrefix.length);
  if (body.length != 48) return false;
  final Uint8List raw;
  try {
    raw = fromHex(body);
  } on FormatException {
    return false;
  }
  final payload = raw.sublist(0, 20), checksum = raw.sublist(20);
  final want = sha256([...utf8.encode(_addressTag), ...payload]).sublist(0, 4);
  return toHex(want) == toHex(checksum);
}

// ------------------------------------------------------------- encryption

Future<String> encryptionPublicFromPrivate(String privHex) async {
  final kp = await _x.newKeyPairFromSeed(fromHex(privHex));
  return toHex((await kp.extractPublicKey()).bytes);
}

Future<Uint8List> _derive(List<int> shared, List<int> epk, List<int> rpk) async {
  final hkdf = Hkdf(hmac: Hmac.sha256(), outputLength: 32);
  final k = await hkdf.deriveKey(
      secretKey: SecretKey(shared), nonce: [...epk, ...rpk], info: utf8.encode(_ecies));
  return Uint8List.fromList(await k.extractBytes());
}

final _aead = Chacha20.poly1305Aead();

Future<Uint8List> _aeadEncrypt(List<int> key, List<int> nonce, List<int> plaintext, List<int> aad) async {
  final box = await _aead.encrypt(plaintext, secretKey: SecretKey(key), nonce: nonce, aad: aad);
  return Uint8List.fromList([...box.cipherText, ...box.mac.bytes]);
}

Future<Uint8List> _aeadDecrypt(List<int> key, List<int> nonce, List<int> ctAndTag, List<int> aad) async {
  if (ctAndTag.length < 16) throw const FormatException('ciphertext too short');
  final ct = ctAndTag.sublist(0, ctAndTag.length - 16);
  final mac = Mac(ctAndTag.sublist(ctAndTag.length - 16));
  final pt = await _aead.decrypt(SecretBox(ct, nonce: nonce, mac: mac), secretKey: SecretKey(key), aad: aad);
  return Uint8List.fromList(pt);
}

/// ECIES: seal [plaintext] so only the holder of [recipientPubHex] can read it.
Future<Map<String, String>> wrapToRecipient(String recipientPubHex, List<int> plaintext) async {
  final rpk = fromHex(recipientPubHex);
  final ephSeed = randomBytes(32);
  final eph = await _x.newKeyPairFromSeed(ephSeed);
  final epk = (await eph.extractPublicKey()).bytes;
  final shared = await _x.sharedSecretKey(
      keyPair: eph, remotePublicKey: SimplePublicKey(rpk, type: KeyPairType.x25519));
  final key = await _derive(await shared.extractBytes(), epk, rpk);
  final nonce = randomBytes(12);
  final ct = await _aeadEncrypt(key, nonce, plaintext, epk);
  return {'epk': toHex(epk), 'nonce': toHex(nonce), 'ct': toHex(ct)};
}

Future<Uint8List> unwrapFromSender(String recipientPrivHex, Map<String, dynamic> blob) async {
  final kp = await _x.newKeyPairFromSeed(fromHex(recipientPrivHex));
  final rpk = (await kp.extractPublicKey()).bytes;
  final epk = fromHex(blob['epk'] as String);
  final shared = await _x.sharedSecretKey(
      keyPair: kp, remotePublicKey: SimplePublicKey(epk, type: KeyPairType.x25519));
  final key = await _derive(await shared.extractBytes(), epk, rpk);
  return _aeadDecrypt(key, fromHex(blob['nonce'] as String), fromHex(blob['ct'] as String), epk);
}

Uint8List newPacketKey() => randomBytes(32);

Future<Uint8List> encryptPacket(List<int> key, List<int> plaintext) async {
  final nonce = randomBytes(12);
  final ct = await _aeadEncrypt(key, nonce, plaintext, utf8.encode(_packetAad));
  return Uint8List.fromList([...nonce, ...ct]);
}

Future<Uint8List> decryptPacket(List<int> key, List<int> ciphertext) async {
  if (ciphertext.length < 12 + 16) throw const FormatException('ciphertext too short');
  return _aeadDecrypt(key, ciphertext.sublist(0, 12), ciphertext.sublist(12), utf8.encode(_packetAad));
}

String keyCommitment(List<int> key) => toHex(sha256([...utf8.encode(_keyCommit), ...key]));

// ------------------------------------------------------------ wallet seal

const scryptN = 1 << 15, scryptR = 8, scryptP = 1;

Uint8List _scrypt(String passphrase, List<int> salt, int n, int r, int p) {
  final s = pc.Scrypt()..init(pc.ScryptParameters(n, r, p, 32, Uint8List.fromList(salt)));
  final out = Uint8List(32);
  s.deriveKey(Uint8List.fromList(utf8.encode(passphrase)), 0, out, 0);
  return out;
}

class WalletLocked implements Exception {
  final String message;
  WalletLocked(this.message);
  @override
  String toString() => message;
}

/// Seal the wallet's secret fields under a passphrase, exactly the reference
/// wallet's `encrypted` blob: scrypt key, ChaCha20-Poly1305 over the
/// canonical JSON of the secrets, with a fixed associated-data tag.
Future<Map<String, dynamic>> sealSecrets(Map<String, dynamic> secrets, String passphrase) async {
  if (passphrase.isEmpty) throw ArgumentError('passphrase must not be empty');
  final salt = randomBytes(16), nonce = randomBytes(12);
  final key = _scrypt(passphrase, salt, scryptN, scryptR, scryptP);
  final ct = await _aeadEncrypt(key, nonce, canonicalBytes(secrets), utf8.encode(_walletAad));
  return {
    'kdf': 'scrypt', 'n': scryptN, 'r': scryptR, 'p': scryptP,
    'salt': toHex(salt), 'nonce': toHex(nonce), 'ct': toHex(ct),
  };
}

Future<Map<String, dynamic>> unsealSecrets(Map<String, dynamic> blob, String passphrase) async {
  if (blob['kdf'] != 'scrypt') throw WalletLocked('unsupported wallet kdf ${blob['kdf']}');
  final key = _scrypt(passphrase, fromHex(blob['salt'] as String), blob['n'] as int, blob['r'] as int, blob['p'] as int);
  try {
    final pt = await _aeadDecrypt(key, fromHex(blob['nonce'] as String), fromHex(blob['ct'] as String), utf8.encode(_walletAad));
    return jsonDecode(utf8.decode(pt)) as Map<String, dynamic>;
  } on SecretBoxAuthenticationError {
    throw WalletLocked('wrong passphrase');
  }
}
